### PR TITLE
feat(dataset): add pagination to the body

### PR DESCRIPTION
**Pull request by ramfox**
This pr adds pagination to the body.

To keep it simple, for now, we are using a `LoadMore` component that has 3 buttons, each adding 100, 500, or 1000 rows respectfully.

when you get to the bottom of a dataset (tabular or not) that has not yet `fetchedAll`, the `LoadMore` component will show.

Body data has been moved out of pagination, and out of entities (more on that later) and into it's own `body` section of the state tree.

The body section contains an `id` (the bodypath), `selector` (stub for the future when we want to filter results), `fetchedAll`, `pageSize`, `error`, and `data` (the actual body data).

We've added a new body reducer and some body actions!

The reducer has 4 options:
- CLEAR_BODY, which returns the initial state
- BODY_REQUEST, which, if the bodypaths or the selectors don't match, returns an inital state with the new bodypath, selector, and pageSize, or if it is the same bodypath and selector, updates the state tree if the pageSize has changed. Either way, `loading` is set to true
- BODY_SUCCESS, which calculates if we have `fetchedAll`, appends the new  body data to any current body data, and sets `loading` to false
- BODY_FAILURE, which sets the `error` and sets `loading` to false

There are tests covering the body reducer

Some actions and selectors have changed!

The `selectDatasetBody` selector now checks to see if the `bodypath` given is the `bodypath` in `state.body.id`, before returning the body.

The actions `loadDatasetBody` and `fetchDatasetBody` actions have been moved to the `actions/body.js` file, and renamed to `loadBodyByPage`, and `fetchBodyByPage`.

We've added `loadBody` and `fetchBody`, that use `offset` and `limit`. This is the main action we use to load the body. We also have `clearBody`, which resets the `state.body`, and we call whenever we navigate away from the dataset page.

In order to use the callApi in our actions, but not have body end up in the `state.entities` tree, we wrote a very light piece of middleware that sits after the callApi, and callCaf middleware that moves the body data out of band if the `action.type` is `BODY_SUCCESS`.

### CODE REVIEW

File: lib/actions/body.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+      schema: Schemas.STRUCTURED_DATA,
+      data: { page, pageSize, format: 'json' }
+    },
+    bodypath,
+    page,
+    pageSize
+  }
+}
+
+export function loadBodyByPage (path, bodypath, page = 1, pageSize = 100) {
```
Comment (b5) [inline lib/actions/body.js]: now that we have limit / offset on the API, let's just remove this entirely until we decide we need it

File: lib/components/dataset/Dataset.js [hunk 2]
```diff
@@ -303,9 +307,17 @@ export default class Dataset extends React.PureComponent {
...
           registryVersion={registryVersion}
 
+          body={body}
+          bodypath={bodypath}
+          noBody={noBody}
+          bodyError={bodyError}
+          bodyFetchedAll={bodyFetchedAll}
+          bodyLoading={bodyLoading}
+          bodyLength={bodyLength}
+          loadBody={loadBody}
```
Comment (b5) [inline lib/components/dataset/Dataset.js:319]: seeing this many `body`-prefixed variables is a good clue that we should group them together into a structure for passing around. That kind of change is perfect as a later refactoring PR

File: lib/reducers/__tests__/body.test.js [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+import { Reducer } from 'redux-testkit'
+import BodyReducer, { initialState } from '../body'
+import {
+  CLEAR_BODY,
+  BODY_REQUEST,
+  BODY_SUCCESS,
+  BODY_FAILURE
+} from '../../constants/body'
+
+describe('Body Reducer', () => {
```
Comment (b5) [inline lib/reducers/__tests__/body.test.js:11]: LET'S GO TESTS

File: lib/middleware/extractBody.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import { BODY_SUCCESS } from '../constants/body'
+
+export default store => next => action => {
+  if (action.type === BODY_SUCCESS) {
```
Comment (b5) [inline lib/middleware/extractBody.js]: when a function only does the main part of it's work if a condition is true, it's easier to read if the negated condition is returned early. eg:
```js

export default store => next => action => {
  if (action.type !== BODY_SUCCESS) {
    return next(action)
  } 

  // ...
}
```

cuts down on lines between brackets and tells the reader that only one situation fires the function. (I totally led you astray while we were working on this, my b)

File: package.json [hunk 5]
```diff
@@ -149,88 +149,88 @@
     }
   },
   "dependencies": {
-    "@babel/core": "7.4.3",
-    "@babel/plugin-proposal-class-properties": "7.4.0",
-    "@babel/plugin-proposal-decorators": "7.4.0",
+    "@babel/core": "7.4.4",
```
Comment (b5) [inline package.json:152]: in the future I'd like to get in the habit of doing any dependency changes as separate PR's

**Review by b5 — APPROVED** [review]
PUMPED. this is looking really good. a few nits. otherwise, LGTM

File: lib/components/dataset/Dataset.js (hunk 2 above)
Comment (ramfox) [reply lib/components/dataset/Dataset.js:319]: oooo v interesting!! noted.

**Pull request merged**